src/powerset.rs

```diff
@@ -0,0 +1,149 @@
+    Empty_,
+}
+
+impl<I> Iterator for Powerset<I>
+    where
+        I: Iterator,
+        I::Item: Clone,
+{
+    type Item = Vec<I::Item>;
+
+    fn next(&mut self) -> Option<Vec<I::Item>> {
+        let result = match &mut self.inner {
+            Inner::Buffer(src, buf) => {
+                // First element is always the empty set.
+                if self.pos == 0 {
+                    Some(alloc::vec![])
+                } else if let Some(elt) = src.next() {
+                    // Return source item as subset with length of one.
+                    buf.push(elt.clone());
+                    Some(alloc::vec![elt])
+                } else {
+                    // Source iterator is exhausted.
+                    if buf.len() > 1 {
+                        // Begin yielding combinations of buffered elements.
+                        self.upgrade_and_get_next()
+                    } else {
+                        return None;
+                    }
+                }
+            },
+            // Iterating through subset combinations.
+            Inner::Combs(combs) => {
+                if let Some(elt) = combs.next() {
+                    Some(elt)
+                } else {
+                    // Try increasing combination length if not already maxed out.
+                    let maxed = combs.k() >= combs.n();
+                    if !maxed && combs.reset(combs.k() + 1) {
+                        combs.next()
+                    } else {
+                        return None;
+                    }
+                }
+            },
+            Inner::Empty_ => { unreachable!() },
+        };
+
+        debug_assert!(result.is_some());
+        self.pos = self.pos.saturating_add(1);
+        result
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        // Total bounds for source iterator (or return 'empty' hint if we're done).
+        let src_total = match &self.inner {
+            Inner::Buffer(src, buf) => size_hint::add_scalar(src.size_hint(), buf.len()),
+            Inner::Combs(combs) => size_hint::add_scalar(combs.src().size_hint(), combs.n()),
+            Inner::Empty_ => unreachable!(),
+        };
+
+        // Total bounds for self ( length(powerset(set) == 2 ^ length(set) )
+        let self_total = size_hint::two_exp(src_total);
+
+        if self.pos < usize::MAX {
+            // Subtract count of already yielded elements from total.
+            size_hint::sub_scalar(self_total, self.pos)
+        } else {
+            // Fallback: self.pos is saturated and no longer reliable.
+            (0, self_total.1)
+        }
+    }
+}
+
+impl<I> Powerset<I>
+    where I: Iterator,
+          I::Item: Clone,
+{
+    // Upgrades the inner state to Inner::Combinations and returns the first
+    // combination, should only be called when self.inner value is Inner::Buffer.
+    #[inline]
+    fn upgrade_and_get_next(&mut self) -> Option<Vec<I::Item>> {
```

Comment:
If we really need this, I think this function should be inlined, as it makes clear that it is only called for `Inner::Buffer`, possibly allowing further simplifications.

Reply:
Originally it was inline in the impl of `Iterator::next()` but moved here to reduce indentation and improve readability.  We can scrap this anyway if we just use the increasing `Combinations` approach.